Sources/SwiftJWT/JWTEncoder.swift

```diff
@@ -0,0 +1,219 @@
+
+// MARK: JWTEncoder
+
+/**
+ A thread safe encoder that signs the JWT header and claims using the provided algorithm and encodes a `JWT` instance as either Data or a JWT String.
+ 
+ ### Usage Example: ###
+ ```swift
+ struct MyClaims: Claims {
+    var name: String
+ }
+ var jwt = JWT(claims: MyClaims(name: "John Doe"))
+ let rsaPrivateKey = read(fileName: "rsa_private_key")
+ let rsaJWTEncoder = JWTEncoder(algorithm: Algorithm.rs256(rsaPrivateKey, .privateKey))
+ do {
+    let jwtString = try rsaJWTEncoder.encodeToString(jwt)
+ } catch {
+    print("Failed to encode JWT: \(error)")
+ }
+ ```
+ */
+public class JWTEncoder: BodyEncoder {
+    
+    let keyIDToSigner: (String) -> JWTSigner?
+    let jwtSigner: JWTSigner?
+    let header: Header?
+    
+    // MARK: Initializers
+    
+    /// Initialize a `JWTEncoder` instance with a single `JWTSigner`.
+    ///
+    /// - Parameter algorithm: The `Algorithm` that will be used to sign the JWT.
+    /// - Returns: A new instance of `JWTEncoder`.
+    public init(jwtSigner: JWTSigner, header: Header? = nil) {
+        self.keyIDToSigner = {_ in return jwtSigner }
+        self.jwtSigner = jwtSigner
+        self.header = header
+    }
+    
+    /// Initialize a `JWTEncoder` instance with a function to generate the `JWTSigner` from the JWT `kid` header.
+    ///
+    /// - Parameter algorithm: The `Algorithm` that will be used to sign the JWT.
+    /// - Returns: A new instance of `JWTEncoder`.
+    public init(keyIDToSigner: @escaping (String) -> JWTSigner?, header: Header? = nil) {
+        self.keyIDToSigner = keyIDToSigner
+        self.header = header
+        self.jwtSigner = nil
+    }
+    
+    // MARK: Encode
+    
+    /// Encode a `JWT` instance into a utf8 encoded JWT String.
+    ///
+    /// - Parameter value: The JWT instance to be encoded as Data.
+    /// - Returns: The utf8 encoded JWT String.
+    /// - throws: `JWTError.invalidUTF8Data` if the provided Data can't be decoded to a String.
+    /// - throws: `JWTError.invalidKeyID` if the KeyID `kid` header fails to generate a jwtSigner.
+    /// - throws: `EncodingError` if the encoder fails to encode the object as Data.
+    public func encode<T : Encodable>(_ value: T) throws -> Data {
+        guard let jwt = try self.encodeToString(value).data(using: .utf8) else {
+            throw JWTError.invalidUTF8Data
+        }
+        return jwt
+    }
+    
+    /// Encode a `JWT` instance as a JWT String.
+    ///
+    /// - Parameter value: The JWT instance to be encoded as a JWT String.
+    /// - Returns: A JWT String.
+    /// - throws: `JWTError.invalidKeyID` if the KeyID `kid` header fails to generate a jwtSigner.
+    /// - throws: `EncodingError` if the encoder fails to encode the object as Data.
+    public func encodeToString<T : Encodable>(_ value: T) throws -> String {
+        let encoder = _JWTEncoder()
+        try value.encode(to: encoder)
+        var _header: Header
+        if let header = header {
```

Comment:
Here we are treating the 'default' header as an override rather than a fallback:  if the encoder was used to encode a JWT (complete with Header), but also had a default header, that one would be used instead.  This seems like it could cause surprises.